src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/OpenSslHelpers.cs

```diff
@@ -86,32 +86,38 @@ public static AlgorithmIdentifier ReadAlgoIdFromEncryptedContentInfo(this DerSeq
                 "Expected to skip an OID while reading EncryptedContentInfo");
             encryptedContentInfo.SkipValue();
 
-            return encryptedContentInfo.ReadAlgoIdentifier();
-        }
+            byte[] contentEncryptionAlgorithmIdentifier = encryptedContentInfo.ReadNextEncodedValue();
 
-        public static AlgorithmIdentifier ReadAlgoIdentifier(this DerSequenceReader encryptedContentInfo)
-        {
-            // The encoding for a ContentEncryptionAlgorithmIdentifier is just a sequence of the OID and the parameters, 
-            // but we just need the OID 
-            DerSequenceReader contentEncryptionAlgorithmIdentifier = encryptedContentInfo.ReadSequence();
-            string algoOid = contentEncryptionAlgorithmIdentifier.ReadOidAsString();
-            int keyLength = 0;
+            DerSequenceReader encryptionAlgorithmReader = new DerSequenceReader(contentEncryptionAlgorithmIdentifier);
+            string algoOid = encryptionAlgorithmReader.ReadOidAsString();
 
-            // TODO(3334): Get the Key length from the algorithm parameters.
-            // The hardcoded values for RC2 and RC4 are not right. Will update accordingly when we work out how to
-            // get them from the parameters. 
+            int keyLength = 0;
             switch (algoOid)
             {
                 case Oids.Rc2:
-                    keyLength = KeyLengths.Rc2_128Bit;
+                    keyLength = Interop.Crypto.CmsGetAlgorithmKeyLength(
+                        contentEncryptionAlgorithmIdentifier, contentEncryptionAlgorithmIdentifier.Length);
+
+                    if (keyLength == -2)
+                    {
+                        System.Diagnostics.Debug.Fail("Call to the shim recieved unexpected invalid input.");
+                        throw new ArgumentNullException();
+                    }
+
+                    if (keyLength == -1)
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+
                     break;
+                    
                 case Oids.Rc4:
-                    int saltLength = 8;
-                    keyLength = KeyLengths.Rc4Max_128Bit - 8 * saltLength;
+                    // We should also set this to the right value but OpenSSL throws an exception when trying to set the ASN1 parameters
+                    // to extract the key length as documented on active issue 10311.
```

Comment:
I'd leave off "active" in "active issue".  Someone could be looking at this code long after the issue gets fixed.